pytensor/tensor/rewriting/elemwise.py

```diff
@@ -529,333 +539,486 @@ def local_upcast_elemwise_constant_inputs(fgraph, node):
+
+        def find_next_fuseable_subgraph(
+            fg: FunctionGraph,
+        ) -> Generator[Tuple[List[Variable], List[Variable]], None, None]:
+            """Find all subgraphs in a FunctionGraph that can be fused together
+
+            Yields
+            -------
+            List of inputs and outputs that determine subgraphs which can be fused. This
+            method assumes that such replacement is done across iterations of the
+            generator.
+            """
+
+            @lru_cache(maxsize=None)
+            def elemwise_scalar_op_has_c_code(node: Apply) -> bool:
+                # TODO: This should not play a role in non-c backends!
+                if node.op.scalar_op.supports_c_code(node.inputs, node.outputs):
+                    return True
+                else:
+                    warn(
+                        (
+                            "Optimization Warning: "
+                            f"The Op {node.op.scalar_op} does not provide a C implementation."
+                            " As well as being potentially slow, this also disables "
+                            "loop fusion."
+                        )
+                    )
+                    return False
+
+            # We start by creating two maps, 1) from each node to each potentially
+            # fuseable client (both nodes must be single output Elemwise with same
+            # broadcast type) and 2) from each node to each certainly unfuseable
+            # client (those that don't fit into 1))
+            fuseable_clients: Dict[Any, List[Any]] = defaultdict(list)
+            unfuseable_clients: Dict[Any, List[Any]] = defaultdict(list)
+            for out, clients in fg.clients.items():
+                out_maybe_fuseable = (
+                    out.owner
+                    and isinstance(out.owner.op, Elemwise)
+                    # and not isinstance(out.owner.op.scalar_op, aes.Composite)
+                    and len(out.owner.outputs) == 1
+                    and elemwise_scalar_op_has_c_code(out.owner)
+                )
+                for client, _ in clients:
+                    if (
+                        out_maybe_fuseable
+                        and not isinstance(client, str)  # "output"
+                        and isinstance(client.op, Elemwise)
+                        # and not isinstance(client.op.scalar_op, aes.Composite)
+                        and len(client.outputs) == 1
+                        and out.type.broadcastable
+                        == client.outputs[0].type.broadcastable
+                        and elemwise_scalar_op_has_c_code(client)
+                    ):
+                        if client not in fuseable_clients[out]:
+                            fuseable_clients[out].append(client)
+                    else:
+                        if client not in unfuseable_clients[out]:
+                            unfuseable_clients[out].append(client)
```

Comment:
Could this work with sets?

Reply:
Don't know if you mean one in specific. I used dictionaries everywhere to maintain rewrite determinism as we populate queues from the dictionary mappings

Reply:
The `if a not in b: b.add(a)` pattern can be a one-liner when using sets, but I don't know if you need these things to be sorted

Reply:
So the `unfuseable_clients` can be a set, because an unfuseable client can never become a fuseable one. But this makes it a bit harder to read because for one I would have to add stuff with `append` and the other with `add`. Also this is the only case where we would save the `not in`, and this is only done once during initialization of the rewrite anyway.